File: neet_code_150_cpp/appendix/sort.cpp
#include <iostream>
#include <algorithm>
#include <vector>

template<typename T>
void show_result(std::vector<T>& result) {
    for(auto& v: result) {
        std::cout << v << " ";
    }
    std::cout << std::endl;
}

int main(void) {
    // 普通のsort
    auto nums = std::vector<int>{5, 18, 2, 10, 8};
    std::sort(nums.begin(), nums.end());
    show_result(nums);

    // 文字列の長さでsort
    auto strings = std::vector<std::string>{"ffoog", "ho", "a", "hogex_xhoge", "abcd", "xyz"};
    // 第三引数にはbool値を返す関数もしくはラムダ式を入れる
    std::sort(strings.begin(), strings.end(), [](auto& s_1, auto& s_2) { return s_1.size() < s_2.size();});
    show_result(strings);
}

File: neet_code_150/src/appendix/sort.rs
fn main() {
    // 普通のsort()
    let mut nums = vec![1, 5, 3, 7, 10, 2];
    nums.sort();
    println!("nums: {:?}", nums);

    // sort_by()で文字列の長さ順でsort
    // sort_by()の引数には配列の型がTのときT&の引数が二個でstd::cmp::Orderingを返すクロージャー
    let mut strings = vec!["ffoog", "ho", "a", "hogex_xhoge", "abcd", "xyz"];
    strings.sort_by(|x, y| x.len().cmp(&y.len()));
    println!("string: {:?}", strings);
}
